### PR TITLE
[SMTChecker] Inline external function calls to `this`

### LINKED ISSUE #6400: [SMTChecker] Inline external function calls to this
Inlining general external function calls is not safe, but for `this.` it is. Those can therefore be inlined.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes #6400

### CODE REVIEW

**Comment by chriseth** [comment]
Please rebase.

File: Changelog.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ Compiler Features:
  * SMTChecker: Support inherited state variables.
  * SMTChecker: Support tuple assignments and function calls with multiple return values.
  * SMTChecker: Support ``delete``.
+ * SMTChecker: Inline external function calls to ``this``.
```
Comment (Marenz) [inline Changelog.md:11]: Maybe add `Support` like the other sentences

**Review by Marenz — APPROVED** [review]

**Review by Marenz — CHANGES_REQUESTED** [review]
LGTM but tests fail

File: Changelog.md (hunk 1 above)
Comment (leonardoalt) [reply Changelog.md:11]: I think support is a different thing, meaning something that is standard. `Inline (...)` is more like a behavior choice

File: libsolidity/formal/SMTChecker.cpp [hunk 2]
```diff
@@ -1901,7 +1904,16 @@ FunctionDefinition const* SMTChecker::inlinedFunctionCallToDefinition(FunctionCa
...
 
 	FunctionType const& funType = dynamic_cast<FunctionType const&>(*_funCall.expression().annotation().type);
-	if (funType.kind() != FunctionType::Kind::Internal)
+	if (funType.kind() == FunctionType::Kind::External)
+	{
+		auto memberAccess = dynamic_cast<MemberAccess const*>(&_funCall.expression());
+		auto identifier = memberAccess ?
+			dynamic_cast<Identifier const*>(&memberAccess->expression()) :
+			nullptr;
+		if (!identifier || identifier->name() != "this")
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]: As we learned recently, the following code is legal:
```
SomeOtherContract this = bla();
this.f();
```
So I think you also have to check the type of `identifier->referencedDeclaration()`.

Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp]: Oh, true...
Even if the type is `ContractDefintion`, can we ever be sure it's the `this` we think it is?
```
contract C
{
    uint public x;
    C c;
    function f(C _c) public {
        c = _c;
    }
    function g() public {
        C this = c;
        this.h();
    }
    function h() public {
        x = 2;
    }
}
```

Comment (chriseth) [reply libsolidity/formal/SMTChecker.cpp]: I think checking for MagicVariableDeclaration should be enough.

**Comment by leonardoalt** [comment]
@chriseth ready for review

**Comment by chriseth** [comment]
Needs rebase.

**Comment by leonardoalt** [comment]
Rebased

**Review by chriseth — APPROVED** [review]

**Pull request merged**